src/duplicates_impl.rs

```diff
@@ -0,0 +1,197 @@
+{
+    fn next_back(&mut self) -> Option<Self::Item> {
+        while let Some(v) = self.iter.next_back() {
+            let key = (self.f)(&v);
+            match self.used.get_mut(&key) {
+                None => {
+                    self.used.insert(key, false);
+                    self.pending += 1;
+                },
+                Some(true) => (),
+                Some(produced) => {
+                    *produced = true;
+                    self.pending -= 1;
+                    return Some(v);
+                },
+            }
+        }
+        None
+    }
+}
+
+impl<I> Iterator for Duplicates<I>
+    where I: Iterator,
+          I::Item: Eq + Hash
+{
+    type Item = I::Item;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        while let Some(v) = self.iter.iter.next() {
+            match self.iter.used.get_mut(&v) {
+                None => {
+                    self.iter.used.insert(v, false);
+                    self.iter.pending += 1;
+                },
+                Some(true) => (),
+                Some(produced) => {
+                    *produced = true;
+                    self.iter.pending -= 1;
+                    return Some(v);
+                },
+            }
+        }
+        None
+    }
+
+    #[inline]
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        let (_, hi) = self.iter.iter.size_hint();
+        // There are `hi` number of items left in the base iterator. In the best case scenario,
+        // these items are exactly the same as the ones pending (i.e items seen exactly once so
+        // far), plus (hi - pending) / 2 pairs of never seen before items.
+        let hi = hi.map(|hi| {
+            let max_pending = min(self.iter.pending, hi);
+            let max_new = max(hi - self.iter.pending, 0) / 2;
+            max_pending + max_new
+        });
+        // The lower bound is always 0 since we might only get unique items from now on
+        (0, hi)
+    }
+}
+
+impl<I> DoubleEndedIterator for Duplicates<I>
+    where I: DoubleEndedIterator,
+          I::Item: Eq + Hash
+{
+    fn next_back(&mut self) -> Option<Self::Item> {
+        while let Some(v) = self.iter.iter.next_back() {
+            match self.iter.used.get_mut(&v) {
+                None => {
+                    self.iter.used.insert(v, false);
+                    self.iter.pending += 1;
+                },
+                Some(true) => (),
+                Some(produced) => {
+                    *produced = true;
+                    self.iter.pending -= 1;
+                    return Some(v);
+                },
+            }
+        }
+        None
+    }
+}
+
+/// An iterator adapter to filter out duplicate elements.
+///
+/// See [`.duplicates()`](../trait.Itertools.html#method.duplicates) for more information.
+#[derive(Clone)]
+#[must_use = "iterator adaptors are lazy and do nothing unless consumed"]
+pub struct Duplicates<I: Iterator> {
+    iter: DuplicatesBy<I, I::Item, ()>,
+}
```

Comment:
I think this could be simplified as:
```suggestion
type Duplicates<I> = DuplicatesBy<I, I::Item, ()>;
```
@phimuemue, do you have any opinions on this sort of simplification? IIRC, the only real danger in it is if we ever want the implementations of `Duplicates` and `DuplicatesBy` to diverge, it would be a breaking change — but that seems unlikely to me.

Reply:
In case this helps resolve the question, I followed the prior art from here https://github.com/rust-itertools/itertools/blob/v0.10.0/src/unique_impl.rs#L139-L146

Reply:
A few recent simplifications we've merged have taken the type alias approach. Let's go with that here, too.

Reply:
@jswrenn I think your simplification idea is a good one if it is doable.

I actually remember that I wanted to do a similar thing for `Unique`/`UniqueBy`, but maybe I hit some obstacles.

Reply:
I tried this today and I also hit some obstacles. First, the type alias can't set the `F` generic parameter to `()` because the implementation of `Iterator` on `DuplicatesBy` has the following trait bounds:

```rust
impl<I, V, F> Iterator for DuplicatesBy<I, V, F>
    where I: Iterator,
          V: Eq + Hash,
          F: FnMut(&I::Item) -> V
```

and I can't provide a separate special implementation for the `F=()` case since Rust doesn't have specialization (yet).

The next best thing would be to have `duplicates` return an instance of `DuplicatesBy` with the closure set to the identity function, but this problematic again because the closure is given a shared reference to the item and has to produce an owned value. This would bring back the `Clone` trait bound on the items and be more wasteful in the common case.

It's been a while since I first worked on this but I now have a faint memory of trying to do that from the beginning and hitting the specialization obstacle. 

Unless there is another way I missed I would keep the code as-is to have good better performance if someone just wants duplicates without a key function.

Reply:
@petrosagg you can unify the implementations like this: https://github.com/jswrenn/rust-itertools/blob/1582e814047560856a767cf280b7617cca6c4736/src/duplicates_impl.rs

Reply:
@jswrenn that is certainly interesting and educational, I have some reading on HRTB to do :) 

The verbose implementation feels way more approachable to me and in the end it's not that much longer, HRTB are a bit exotic. But feel free to merge whichever version you think is best for itertools.

Reply:
I'd like to go forward with the unified implementation. Here's why: 

The trickiest part of writing iterators is the methods. The implementations of `next`, `next_back` and any overridden methods tend to be very subtle, and each additional method implemented/modified requires careful testing. For the lifetime of this iterator, we probably _won't_ get any PRs modifying the `Key` abstraction, but we probably _will_ get PRs for optimizing `next`, `next_back` and adding additional overrides. 

Unifying these two similar implementations halves the amount of subtle code that these PRs will have to account for. It'll make those improvements easier to implement, test, and review.

Reply:
No problem at all. Do you mind giving a short explanation on how it works? I've been trying to understand but I don't get it. 

In particular, in the `ById` case, how can the `key` produced by `self.k.key(&v);` be inserted into the `used` HashMap without lifetime issues? From what I understand a reference to the just produced item is inserted and not the item itself, but then the item surely goes out of scope? What am I missing?

Reply:
Whoops, you're right: that _doesn't_ work! (Invocations of `duplicates` fail with a compiler error.)

Reply:
hm, looking at it a bit more I'm not sure the proposed solution works. If I cargo test I get these kind of errors for invocations of `duplicates()`:

```
error[E0277]: the trait bound `for<'i> itertools::duplicates_impl::private::ById: itertools::duplicates_impl::private::Key<'i, i8, i8>` is not satisfied
   --> tests/quick.rs:920:27
    |
287 | fn correct_size_hint<I: Iterator>(mut it: I) -> bool {
    |                         -------- required by this bound in `correct_size_hint`
...
920 |         correct_size_hint(it.duplicates())
    |                           ^^^^^^^^^^^^^^^ the trait `for<'i> itertools::duplicates_impl::private::Key<'i, i8, i8>` is not implemented for `itertools::duplicates_impl::private::ById`
    |
    = help: the following implementations were found:
              <itertools::duplicates_impl::private::ById as itertools::duplicates_impl::private::Key<'i, Item, &'i Item>>
    = note: required because of the requirements on the impl of `Iterator` for `itertools::duplicates_impl::private::DuplicatesBy<Iter<i8>, i8, itertools::duplicates_impl::private::ById>`

error: aborting due to previous error
```

Which matches the intuition I described above, the call tries to use a `for<'i> Key<'i, i8, i8>` but only an implementation of `for<'i> Key<'i, i8, &'i i8>` exists (which in turn is not suitable because we need to store the `Output` into the HashMap. Am I missing something?

Reply:
Nope! References definitely pose the challenge here.

I wonder if dropping the `Clone` bound is worth the ergonomic cost. E.g., this is the sort of situation in which I would use `duplicates_by` instead of `duplicates`:
```rust
struct Foo {
    bar: String,
    baz: u8,
}

let foos = 
    ('a'..'z').chain('a'..'z').zip(0..)
        .map(|(bar, baz)| Foo {bar: bar.to_string(), baz});

let dups = duplicates_by(foos, |foo| foo.bar); // Compile error!
```
...but that's a compile error.

Reply:
@jswrenn I cracked the puzzle! I updated the PR which now has a single implementation that is shared not only between `Duplicates` and `DuplicatesBy` but also between the `Iterator` and `DoubleEndedIterator`.